pkg/log/logger.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright 2017 Mirantis
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package log
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+	"unicode"
+
+	"github.com/hpcloud/tail"
+)
+
+type VirtletLogger interface {
+	SpawnWorkers() error
+	StopObsoleteWorkers()
+}
+
+type WorkerRunner interface {
+	RunNewWorker(inputFile, outputFile, sandboxId string)
+	ListWorkers() []string
+	WorkerExists(sandboxId string) bool
+	StopWorker(sandboxId string)
+	StopAllWorkers()
+	InitStatusChannel() chan string
+}
+
+var _ VirtletLogger = (*virtletLogger)(nil)
+var _ WorkerRunner = (*workerRunner)(nil)
+
+func NewVirtletLogger(virtletFolder, kubernetesFolder string) VirtletLogger {
+	logger := &virtletLogger{
+		config: virtletLoggerConf{
+			VirtletFolder:      virtletFolder,
+			VirtletFilename:    "raw.log",
+			KubernetesFolder:   kubernetesFolder,
+			KubernetesFilename: "_0.log",
+		},
+		workerRunner: newWorkerRunner(),
+	}
+	return logger
+}
+
+type virtletLoggerConf struct {
+	VirtletFolder      string
+	VirtletFilename    string
+	KubernetesFolder   string
+	KubernetesFilename string
+}
+
+type virtletLogger struct {
+	config       virtletLoggerConf
+	workerRunner WorkerRunner
+}
+
+func (v *virtletLogger) SpawnWorkers() error {
+	fmt.Println("Check and spawn workers")
+
+	vmFolders, err := ioutil.ReadDir(v.config.VirtletFolder)
+	if err != nil {
+		return err
+	}
+
+	for _, vmFolder := range vmFolders {
+		sandboxId := vmFolder.Name()
+
+		if v.workerRunner.WorkerExists(sandboxId) {
+			fmt.Printf("worker for sandbox '%s' already running. Skip.\n", sandboxId)
```

Comment:
If error message goes out to the user it should be capitalized.
It should start with lower case only (and always) when it's returned from func as an `error`.